core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/NormalizeOrExpressionRewriter.java

```diff
@@ -0,0 +1,99 @@
+import io.trino.sql.tree.InListExpression;
+import io.trino.sql.tree.InPredicate;
+import io.trino.sql.tree.LogicalExpression;
+
+import java.util.List;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.sql.ExpressionUtils.and;
+import static io.trino.sql.ExpressionUtils.or;
+import static io.trino.sql.tree.ComparisonExpression.Operator.EQUAL;
+import static io.trino.sql.tree.LogicalExpression.Operator.AND;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+
+public final class NormalizeOrExpressionRewriter
+{
+    public static Expression normalizeOrExpression(Metadata metadata, Expression expression)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(metadata), expression);
+    }
+
+    private NormalizeOrExpressionRewriter() {}
+
+    private static class Visitor
+            extends ExpressionRewriter<Void>
+    {
+        private final Metadata metadata;
+
+        public Visitor(Metadata metadata)
+        {
+            this.metadata = requireNonNull(metadata, "metadata is null");
+        }
+
+        @Override
+        public Expression rewriteLogicalExpression(LogicalExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            List<Expression> terms = node.getTerms().stream()
+                    .map(expression -> treeRewriter.rewrite(expression, context))
+                    .collect(toImmutableList());
+
+            if (node.getOperator() == AND) {
+                return and(terms);
+            }
+
+            ImmutableMultimap.Builder<Expression, Expression> builder = ImmutableMultimap.builder();
+            terms.stream()
+                    .filter(NormalizeOrExpressionRewriter::isEqualityComparisonExpression)
+                    .map(ComparisonExpression.class::cast)
+                    .collect(groupingBy(ComparisonExpression::getLeft, mapping(ComparisonExpression::getRight, toImmutableList())))
+                    .forEach((key, value) -> {
+                        if (value.size() > 1) {
+                            builder.putAll(key, value);
+                        }
+                    });
+
+            Multimap<Expression, Expression> expressionToBeTranslated = builder.build();
+
+            ImmutableList.Builder<Expression> termsBuilder = ImmutableList.builder();
+            for (Expression expression : terms) {
+                if (!(isEqualityComparisonExpression(expression) && expressionToBeTranslated.containsKey(((ComparisonExpression) expression).getLeft()))) {
+                    termsBuilder.add(expression);
+                }
+            }
+
+            expressionToBeTranslated.asMap().forEach((left, right) -> termsBuilder.add(new InPredicate(left, new InListExpression(ImmutableList.copyOf(right)))));
+
+            return or(termsBuilder.build());
```

Comment:
This can be done side-effect free and more concisely with:

```java
List<InPredicate> comparisons = terms.stream()
        .filter(NormalizeOrExpressionRewriter::isEqualityComparisonExpression)
        .map(ComparisonExpression.class::cast)
        .collect(groupingBy(
                ComparisonExpression::getLeft,
                collectingAndThen(
                        mapping(ComparisonExpression::getRight, Collectors.toList()),
                        InListExpression::new)))
        .entrySet().stream()
        .map(e -> new InPredicate(e.getKey(), e.getValue()))
        .collect(Collectors.toList());

List<Expression> others = terms.stream()
        .filter(expression -> !isEqualityComparisonExpression(expression))
        .collect(Collectors.toList());

return or(ImmutableList.<Expression>builder()
        .addAll(others)
        .addAll(comparisons)
        .build());
```

Reply:
Thanks for the snippet. Have modified the `map` function so as to skip creating `IN` expression if we have a single `comparison` expression.